configuration.k

```diff
@@ -0,0 +1,220 @@
+                // Needed or not? It is part of KEEI, but not sure. Should discuss. 
+                <block>
+                    <hashes>           .List      </hashes>
+                    <coinbase>         0          </coinbase>         // H_c
+                    // <logsBloom>        .WordStack </logsBloom>        // H_b
+                    <difficulty>       0          </difficulty>       // H_d
+                    <number>           0          </number>           // H_i
+                    <gasLimit>         0          </gasLimit>         // H_l
+                    <timestamp>        0          </timestamp>        // H_s
+                </block>
+            </execEngine>
+            
+            // This is the part of the "Solidity VM" where we keep the available contracts. 
+            // Those are either obtained from source file (creation) or loaded from accounts (function call).
+
+            // When we parse Solidity source file we 
+            //   - desugar/preprocess modifiers, inheritance (?) etc. 
+            //   - create one <contract> structure for each contract declared in the sources
+            // (let's not worry about those for now, and just stick to simple functions)
+
+            // From here, <contract>s can be either deployed or executed. 
+            // When a contract is deployed we not only copy its <contract> cell into the <account>, 
+            // but we also copy the other contracts that were present in the same source file, 
+            // so that the deployed contract have access to them and can create and deploy them if needed. 
+            // (see the <account> cell below for more notes about this)
+
+            <contracts>
+                <Contract multiplicity = "*">
+                    <cname> .K </cname>
+                    
+                    <functions>
+                        <Function multiplicity="*">
+                            <fname>         .K         </fname>
+                            <inputParams>  .K         </inputParams> 
+                            <Returns>      .K         </Returns>
+                            <quantifiers>  .K         </quantifiers>
+                            <visibility>   .K         </visibility> // Visibility ::= public | privat
+                            <body> .K </body>
+                        </Function>
+                    </functions>
+		            
+                    <stateVar> .List </stateVar>
+                    <Constructor> false </Constructor>
+                    // ... more to be added later ...
+                
+                </Contract>
+            </contracts>
+
+        </solidity>
+
+        // Ethereum Network
+        // ==========================
+
+        <network>
+
+            // Accounts Record
+            // ---------------
+
+            <activeAccounts> .Set </activeAccounts>
+	        <cntAccount> 0:Int </cntAccount>
+            <accounts>
+                <account multiplicity="*">
```

Comment:
and this

Reply:
OK, I will revise it accordingly.

Reply:
I just found that we define `<account>` as a "bag" instead of a "map". If I add `type="Map"`, there will be errors as follows:
```[Error] Critical: failed to concatenate maps with common keys:```

Reply:
If I change add `type="Map"` to the `<account>`, even the following example has the same error:
`"code" : #solidity( new C(); )`

And, I found that the problem comes from the following rule:
```
rule [Allocate-State-Variables]:
    <k> #allocate(N, #varInfo(X:Id,E:Value,T:TypeName, #storage)) => . ...</k>
    <account>
        <acctID> #account(N,_:Id) </acctID>
        <context> CONTEXT:Map => CONTEXT[X <- #storedVar(!N:Int, T, #storage)] </context>
        <Storage> STORAGE:Map => STORAGE[!N <- E] </Storage>
        ...
    </account>
```

After applying the above rule, the same error happens:
[Error] Critical: failed to concatenate maps with common keys:

```
[(
    <acctID>(#account(Int(#"0"),, Id(#"C"))),
    <account>(
        <acctID>(#account(Int(#"0"),, Id(#"C"))),,
        <balance>(Int(#"0")),, 
        <context>(Id(#"a") |-> #storedVar(Int(#"1"),,int_SOLIDITY-SYNTAX(.KList),, #storage_SOLIDITY-SYNTAX(.KList))   ),,
        <code>(
            <associatedContract>(#EmptyK()),, 
            <availableContracts>(#EmptyK())),,
        <Storage>(Int(#"1") |-> #undef_Value_SOLIDITY-SYNTAX(.KList)   ),,
        <nonce>(Int(#"0"))),
    
    <account>(
        <acctID>(#account(Int(#"0"),, Id(#"C"))),,
        <balance>(Int(#"0")),, 
        <context>(.Map),,
        <code>(
            <associatedContract>(#EmptyK()),, 
            <availableContracts>(#EmptyK())),,
        <Storage>(.Map),, 
        <nonce>(Int(#"0")))
)]

  while evaluating function _AccountCellMap_
```

Reply:
@shangweilin @dfilaretti  I can't reproduce this. How did you get those results? I'm using the most recent version of K.  `krun tests/functionCall.sol` ends with `<mode> "SUCCESS" </mode>` and the other cells also look OK. BTW how did you get the step number?

Reply:
@tomtomjhj, to reproduce the error, you need to do the following changing of code:
1. Change Line 149 of `configuration.k` as `<account multiplicity="*" type="Map">`
2. Change Line 49 of `functionCall.sol` as
```
"code" : #solidity( new C();)
```

I got the step number by running the following command, where `i` from 1 to 37
`krun test/functionCall.sol --depth i`
The error happens from steps 37 to 38.

Reply:
I looked into the implementation of K builtin Map and kompiled rules. This is how a Map element rewriting is done:
1. look up the key, 
2. remove this key-value pair from the Map
3. update the value of this entry
4. concatenate this updated entry with the Map (without this key) from 2.

The key of the Map is `<acctID>` cell and its contents. In your rule [Allocate-State-Variables], the key is `<acctID> #account(N,_:Id) </acctID>` containing a K variable `_:Id`, which is an entity that has no relationship with any concrete terms before it is unified to something else.
The problem here is that the actual key in the map `<acctID> #account( 0, C) </acctID>` is different from the key your rule looks up for because `_:Id` is not `C`. Therefore, the lookup of the first step fails, no key-value pair removed, then finally, the concatenation of Maps with common key.

To resolve this, I would only put the account Id on the `<acctID>` cell and devise a way to get the contract name when needed instead of using `#account(..)`.
e.g. new `<contractName>` subcell of `<account>`
```
 ...
 <account>
    <acctID> ID </acctID>
    <contractName> C </contractName>
    ...
 </account>
 ...
```

Reply:
Thanks for your information, @tomtomjhj. It really helps to solve the problem. I have revised the rules so that we match the exact/correct key. Now, the problem is solved.

Reply:
Thanks @tomtomjhj!!